WordPress/gutenberg
# Dimensions Panel: Add new ToolsPanel component and update spacing supports

Part of: https://github.com/WordPress/gutenberg/issues/28356

## Description

- Adds new `ToolsPanel` components
- Uses the new components to refactor spacing block supports into a "Dimensions" panel.

The new panel component offers progressive discovery options for `ToolsPanelItem` wrapped components e.g. those provided through block supports. 

It uses callbacks passed as props on the panel's children to check for values, handle child selection/deselection in the menu etc. In the case of block support controls, the `onDeselect` callback can reset the block attribute value. 

The `ToolsPanel` automatically generates its menu from children that register themselves via the panel's context. The initial state managing which controls are displayed and therefore selected in the menu is based off the `hasValue` callback and an `isShownByDefault` prop.

A custom `resetAll` callback is passed to the panel and called when the associated menu item is selected. For the a panel displaying block support controls such as the dimensions panel, the `resetAll` callback will reset all block attributes for the dimensions related support. 

#### G2 Components
The [demo typography panel](https://g2-components.xyz/?path=/story/designtools-presentation-typographypanel--default) from the G2 project still needs a few underlying components and systems to be imported into Gutenberg. The approach in this PR works now and may unblock other block supports work.

New designs/G2 components for the individual block support controls can be iterated on separately.

## How has this been tested?
Programmatic tests are available for the new panel: 
- `npm run test-unit packages/components/src/tools-panel/test/`

Storybook: 
- `npm run storybook:dev`
-  http://localhost:50240/?path=/story/components-toolspanel--default

Manual testing is required for the updates switching the spacing block support UI to the new panel and "Dimensions" naming.

#### Testing Setup

The easiest block to test with is the group block. 

1. Turn on `settings.spacing.customMargin` under your theme.json file.
2. Update the group block's `block.json` and turn on margin support under `supports.spacing`.
3. Add new `__experimentalDefaultControls` flag under `settings.spacing` in block.json to specify which block support controls are displayed by default.

The example below configures only the padding control to show by default.
```json
		"spacing": {
			"margin": true,
			"padding": true,
			"__experimentalDefaultControls": {
				"padding": true
			}
		},
```

#### Testing Instructions

The following instructions assume the padding control was enabled by default as per the snippet above.

1. Edit a post, add a group block and select it
2. Confirm there is a new "Dimensions" panel in the sidebar in place of the old "Spacing" panel
3. Ensure the padding control is already displayed in the Dimensions panel
4. Select margin control for display via the "Dimensions" panel's "more" menu
5. Adjust padding and margin, confirming they update the block appropriately
6. Open the more menu and toggle off the padding control
7. Confirm the padding control is no longer shown
8. Open the more menu once more and toggle off the margin control
9. The margin control should be removed as well
10. Toggle both controls back on and confirm they've been reset
11. Adjust both padding and margin values, then select the "reset all" option from the more menu
12. Confirm only the padding control is displayed but has been cleared
13. Toggling margin back on and adjust both values again, then save
14. Confirm margin and padding are displayed correctly on the frontend
15. Reload the editor and confirm that the "Dimensions" panel has both fields displayed
16. Choose the "Reset all" option from the panel menu again and save
17. Reload both the editor and frontend checking the padding and margin were cleared
18. Open the full site editor, open the Global Styles sidebar, then the "By block type" tab
19. Expand the Group block's panel and confirm the "Dimensions" panel is present with both margin and padding shown by default. 

## Screenshots 
#### With Default Controls
<img src="https://user-images.githubusercontent.com/60436221/120792288-9e72c080-c578-11eb-81ff-1d165e21f185.gif" width="300" />

#### Without Default Controls
<img src="https://user-images.githubusercontent.com/60436221/120450642-fe7c3200-c3d3-11eb-9f61-a578999f84c2.gif" width="500"/>


## Types of changes
New feature.

## Checklist:
